File: web/src/types/contentEvent.ts
export type ContentBlock = { block_id: string; text: string; source_refs: string[] }
export type ContentFrame = {
  schema_version: 'xunke-content.v1'
  task_id: string
  execution_id: string
  generation_revision: number
  seq: number
  type: 'snapshot' | 'delta' | 'block' | 'reset' | 'finalized' | 'unavailable' | 'revoked'
  payload: {
    block_id?: string
    text?: string
    validated_blocks?: ContentBlock[]
    validation?: 'preview' | 'structure_and_references_checked'
    reason?: string
  }
}

export type ContentPreviewState = {
  taskId?: string
  executionId?: string
  generationRevision: number
  seq: number
  text: string
  blocks: ContentBlock[]
  status: 'waiting' | 'streaming' | 'reconnecting' | 'finalized' | 'unavailable' | 'revoked'
}

export const emptyContentPreview: ContentPreviewState = {
  generationRevision: 0, seq: 0, text: '', blocks: [], status: 'waiting',
}

File: backend/app/models/content_event.py
"""Private content frames; deliberately separate from public task phase events."""

from typing import Literal

from pydantic import Field

from app.rag.contracts import Contract


class ContentBlock(Contract):
    block_id: str = Field(min_length=1, max_length=80)
    text: str = Field(min_length=1, max_length=6000)
    source_refs: list[str] = Field(default_factory=list, max_length=20)


class ContentPayload(Contract):
    block_id: str | None = Field(default=None, max_length=80)
    text: str | None = Field(default=None, max_length=6000)
    validated_blocks: list[ContentBlock] | None = Field(default=None, max_length=30)
    validation: Literal["preview", "structure_and_references_checked"] | None = None
    reason: Literal[
        "new_draft", "attempt_changed", "cursor_expired", "stream_unavailable",
        "preview_limit", "generation_failed", "source_revoked", "session_expired",
        "final_result_available", "preview_expired",
    ] | None = None


class ContentFrame(Contract):
    schema_version: Literal["xunke-content.v1"] = "xunke-content.v1"
    task_id: str
    execution_id: str
    generation_revision: int = Field(ge=1)
    seq: int = Field(ge=0)
    type: Literal["snapshot", "delta", "block", "reset", "finalized", "unavailable", "revoked"]
    payload: ContentPayload = Field(default_factory=ContentPayload)
